### PR TITLE
implement schema-format parsers for RFC 68

### PR DESCRIPTION
**Pull request by cdisselkoen**
## Description of changes

Implements the needed changes for RFC 68 in the parsers for both schema formats (Cedar schema format and JSON schema format).  Includes tests that things parse properly and that invalid uses of embedded attribute maps correctly fail to parse.  Does not include any changes to the validator/typechecker itself; those will come in a future PR.  These parser changes are reasonably separable though, and this PR is large enough as-is. 

With this PR, if anyone tries to use the new RFC 68 syntax, instead of an incomprehensible parser error, they will get a nice message saying that embedded attribute maps are not fully supported yet.

## Issue #, if available

Part of #1104 

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A bug fix or other functionality change requiring a patch to `cedar-policy`.

I confirm that this PR (choose one, and delete the other options):

- [x] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)

### CODE REVIEW

**Comment by shaobo-he-aws** [comment]
There are merge conflicts. Could you resolve them?

**Comment by cdisselkoen** [comment]
Working on it

**Comment by cdisselkoen** [comment]
Merge conflicts fixed

**Review by john-h-kastner-aws — APPROVED** [review]

File: cedar-policy-validator/src/cedar_schema/to_json_schema.rs [hunk 1]
```diff
@@ -366,49 +455,84 @@ fn convert_entity_decl(
...
+    match attr.node {
+        AttrDecl::Concrete { name, required, ty } => Ok((
+            name.node,
+            json_schema::EntityAttributeType {
+                ty: cedar_type_to_entity_attr_type(ty)?,
+                required,
+            },
+        )),
+        AttrDecl::EAMap { .. } => {
+            // EAMap is not allowed here, as the descriptor of all entity attributes.
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/cedar_schema/to_json_schema.rs:529]: Done in the latest commit

File: cedar-policy-validator/src/json_schema.rs [hunk 2]
```diff
@@ -986,8 +1249,103 @@ impl TypeFields {
...
+    /// If this is `Some`, the `additional_attributes` field is present, with the given value
+    additional_attributes: Option<std::result::Result<bool, E>>,
+    /// If this is `Some`, the `name` field is present, with the given value
+    name: Option<std::result::Result<SmolStr, E>>,
+    /// If this is `Some`, the `default` field is present, with the given value
+    default: Option<std::result::Result<Type<N>, E>>,
+}
+
+impl<N, E> Default for TypeFieldsWithData<N, E> {
+    fn default() -> Self {
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/json_schema.rs:1281]: The `derive`'d impl requires `N` to itself be `Default`, whereas this manual implementation does not have any restrictions on `N`. I'll add a comment.

File: cedar-policy-validator/src/cedar_schema/to_json_schema.rs [hunk 3]
```diff
@@ -94,24 +96,106 @@ fn is_valid_ext_type(ty: &Id, extensions: &Extensions<'_>) -> bool {
...
+                    // PANIC SAFETY: already determined `ea_map_decls.len()` is at least 2
+                    #[allow(clippy::indexing_slicing)]
+                    let (source_loc_2, _) = ea_map_decls[1];
+                    Err(MultipleEAMapDeclarationsError {
+                        source_loc_1: source_loc_1.clone(),
+                        source_loc_2: source_loc_2.clone(),
+                    }
+                    .into())
+                }
+                (0, 1..) | (1, 2..) => {
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/cedar_schema/to_json_schema.rs]: Great suggestion.  Ended up using `Itertools::partition_map` instead.

**Review by shaobo-he-aws — APPROVED** [review]
LGTM.

**Review by shaobo-he-aws — APPROVED** [review]

**Review by shaobo-he-aws — APPROVED** [review]

File: cedar-policy-validator/src/cedar_schema/test.rs [hunk 4]
```diff
@@ -2275,3 +2316,163 @@ mod common_type_references {
...
+                &ExpectedErrorMessageBuilder::error("error parsing schema: found an embedded attribute map type, but embedded attribute maps are not allowed in this position")
+                    .exactly_one_underline("?: String")
+                    .build(),
+            );
+        });
+    }
+
+    #[test]
+    fn common_type() {
+        let src = "type blah = { ?: String }; entity User { blah: blah };";
```
[the remark this answers is not in the document]
Comment (john-h-kastner-aws) [reply cedar-policy-validator/src/cedar_schema/test.rs:2408]: Sounds fine.

**Review by john-h-kastner-aws — APPROVED** [review]

**Pull request merged**